python/ray/serve/_private/application_state.py

```diff
@@ -141,127 +237,189 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
                 "path in your application to avoid this issue."
             )
 
-        self._update_status(ApplicationStatus.DEPLOYING)
-        return cur_deployments_to_delete
+        for params in deployment_params:
+            params["deployment_name"] = params.pop("name")
+            params["app_name"] = self._name
+
+        deployment_infos = {
+            params["deployment_name"]: deploy_args_to_deployment_info(**params)
+            for params in deployment_params
+        }
+        self._set_target_state(deployment_infos=deployment_infos)
+
+    def update_obj_ref(self, deploy_obj_ref: ObjectRef, deployment_time: int) -> None:
+        """Update deploy object ref.
+
+        Updates the deployment timestamp, sets status to DEPLOYING, and
+        wipes the current target state. While the task to build the
+        application is running, the application state stops the
+        reconciliation loop and waits on the task to finish and call
+        apply_deployments_args().
+        """
+
+        if self._deploy_obj_ref:
+            logger.info(
+                f'Received new config for application "{self._name}". '
+                "Cancelling previous request."
+            )
+            ray.cancel(self._deploy_obj_ref)
```

Comment:
I think there is a race condition here.

Because we do this "fire and forget," there may be a (very rare) situation where we cancel this obj ref and start a new one, but this one actually ends up running and overwriting the goal state. In this case we would have an internal state inconsistency: we think the application is updated but it's actually running a previous version.

This is another reason why we should change the behavior such that the `deploy_obj_ref` _returns_ the goal state and it gets applied here: that way we can reason about and avoid any such race conditions from having multiple writers.

I don't think this is a regression from the current behavior, so I'm ok not fixing it in this PR assuming that you plan to follow up with the fix suggested above before the 2.6 branch cut.

Reply:
Yeah, changing the task to only return the goal state, and checking on it in application state manager will be the last part of the deploy refactor. However since I will be OOO next week, I think it's unlikely I'll be able to get this in before the 2.6 branch cut. Do you think it's acceptable to get this fix in for 2.7?

Reply:
Assuming this is not a regression from existing behavior I'm OK with it going in 2.7